README.md
# InovaAwsS3
InovaAwsS3 is a Ruby gem designed to simplify interactions with Amazon S3, providing seamless integration for uploading, managing, and retrieving files from your S3 buckets. Whether you're building a web application, handling backups, or managing assets, InovaAwsS3 offers a clean and efficient interface to harness the power of Amazon S3 within your Ruby projects.
## Installation

To use this gem in your project, add it to your application's Gemfile:

```ruby
gem 'inova_aws_s3', '~> 0.5.0'
```

Alternatively, you can install the gem directly using:

```bash
gem install inova_aws_s3 -v '~> 0.5.0'
```
## Requirments

- Ruby >= 3.1.2
- Rails >= 6.1.7

## Usage

```bash
rails generate inova_aws_s3:install
```

This will create a configuration file at config/initializers/inova_aws_s3.rb. Open the file and set your AWS credentials:

```ruby
InovaAwsS3.configure do |config|
  config.acl = 'public-read'
  config.bucket_name = 'your_bucket_name'
  config.region = 'your_region'
  config.access_key = 'your_access_key'
  config.secret_key = 'your_secret_key'
  config.folder_name = 'your_folder_name'
end
```

Replace the placeholder values with your actual AWS credentials.

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/ahmedmos3ad/inova_aws_s3/pulls. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [code of conduct](https://github.com/ahmedmos3ad/inova_aws_s3/blob/master/CODE_OF_CONDUCT.md).

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

## Code of Conduct

Everyone interacting in the InovaAwsS3 project's codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/ahmedmos3ad/inova_aws_s3/blob/master/CODE_OF_CONDUCT.md).


CHANGELOG.md
## [0.3.0] - 2023-11-23

- Fix install generator typo

## [0.1.0] - 2023-11-23

- Initial release


test/test_helper.rb
# frozen_string_literal: true

$LOAD_PATH.unshift File.expand_path("../lib", __dir__)
require "inova_aws_s3"

require "minitest/autorun"


test/test_inova_aws_s3.rb
# frozen_string_literal: true

require "test_helper"
require "minitest/autorun"

require "inova_aws_s3"
require_relative "../lib/inova_aws_s3/test_config"

class TestInovaAwsS3 < Minitest::Test
  def setup
    # Initialize the configuration for each test
    InovaAwsS3.configuration = InovaAwsS3::Configuration.new
  end

  def test_that_it_has_a_version_number
    refute_nil ::InovaAwsS3::VERSION
  end

  def test_onepart_upload
    # Mock the necessary objects and test onepart upload functionality
    s3_service = InovaAwsS3::Service.new
    # Set up mock file and other necessary data
    result = s3_service.onepart_upload(file_name: "Gemfile.lock", tempfile: "Gemfile.lock")
    # Assert the result based on your expectations
    assert_match URI::DEFAULT_PARSER.make_regexp, result
  end

  def test_configuration
    InovaAwsS3.configure do |config|
      config.acl = "public-read"
    end

    assert InovaAwsS3.configuration.acl == "public-read"
  end
end


lib/inova_aws_s3/test_config.rb
# test_config.rb

module InovaAwsS3
  class Configuration
    require "dotenv"
    Dotenv.load(".env")
    attr_accessor :acl, :bucket_name, :region, :access_key, :secret_key, :folder_name

    def initialize
      # Set test-specific configuration values here
      @acl = ENV["TEST_ACL"]
      @bucket_name = ENV["TEST_BUCKET"]
      @region = ENV["TEST_REGION"]
      @access_key = ENV["TEST_ACCESS_KEY"]
      @secret_key = ENV["TEST_SECRET_KEY"]
      @folder_name = ENV["TEST_FOLDER"]
    end
  end
end


lib/inova_aws_s3/service.rb
module InovaAwsS3
  class Service
    require "aws-sdk-s3"
    PART_SIZE = 1024 * 1024 * 10

    def initialize(acl: InovaAwsS3.configuration.acl)
      @acl = acl
      @bucket_name = InovaAwsS3.configuration.bucket_name
      @region = InovaAwsS3.configuration.region
      @access_key = InovaAwsS3.configuration.access_key
      @secret_key = InovaAwsS3.configuration.secret_key
      @folder_name = InovaAwsS3.configuration.folder_name
      set_s3_client
      set_s3_resource
    end

    def set_s3_client
      @s3_client = Aws::S3::Client.new(region: @region, access_key_id: @access_key, secret_access_key: @secret_key)
    end

    def set_s3_resource
      @s3_resource = Aws::S3::Resource.new(region: @region, access_key_id: @access_key, secret_access_key: @secret_key)
    end

    def multipart_upload(filename: nil, file: nil, acl: "public-read")
      logger ||= Logger.new("#{Rails.root}/log/uploads.log")
      key = @folder_name + filename
      resp = @s3_client.create_multipart_upload({
        bucket: @bucket_name,
        key: key, acl: acl
      })
      current_part = 1
      parts = []
      file = File.open(file, "rb")
      until file.eof?
        logger.info("#{key} Part #{current_part}")
        part = file.read(PART_SIZE)
        resp_upload = @s3_client.upload_part({
          body: part,
          bucket: @bucket_name,
          key: key,
          part_number: current_part,
          upload_id: resp.upload_id
        })

        parts << {etag: resp_upload.etag, part_number: current_part}
        current_part += 1
      end
      @s3_client.complete_multipart_upload({bucket: @bucket_name, # required
                                            key: key, # required
                                            multipart_upload: {
                                              parts: parts
                                            },
                                            upload_id: resp.upload_id})
      file.close
      set_object(key).public_url
    end

    def onepart_upload(file_name: nil, tempfile: nil)
      file_name = "#{SecureRandom.urlsafe_base64}_#{Time.now.strftime("%s")}_#{File.extname(file_name)}".gsub(/\s+/, "")
      key = @folder_name + file_name
      obj = set_object(key)
      obj.upload_file(tempfile, {acl: @acl})
      obj.public_url
    end

    def set_object(key)
      @s3_resource.bucket(@bucket_name).object(key)
    end

    def change_acl(url: nil, new_acl: "authenticated-read")
      return if url.nil?
      key = get_key_from_url(url)
      @s3_client.put_object_acl({acl: new_acl, bucket: @bucket_name, key: key})
    rescue
      puts "error in #{url}"
    end

    def get_key_from_url(url)
      # uri = URI.parse(url)
      # File.basename(uri.path)
      arr = url.split("/")
      arr[3..]&.join("/")
    end

    def valid_s3_url?(url)
      arr = url.split("/")
      host_name = "#{@bucket_name}.s3.#{@region}.amazonaws.com"
      arr[2] == host_name
    end

    def get_key(filename)
      @folder_name + filename
    end

    def generate_get_presigned(url)
      key = get_key_from_url(url)
      object = @s3_resource.bucket(@bucket_name).object(key)
      object.presigned_url(:get, expires_in: 10800) # 3 hours
    end

    def get_presigned_url(file_name)
      object = @s3_resource.bucket(@bucket_name).object(@folder_name.to_s + file_name)
      presigned_url = object.presigned_url(:put, {acl: "public-read"}) # , expires: 10*60)
      {presigned_url: presigned_url, file_url: object.public_url}
    end

    def delete_file(url)
      key = get_key_from_url(url)
      return if key.blank?
      @s3_client.delete_object({bucket: @bucket_name, key: key})
    end
  end
end


lib/aws/s3_service.rb
module Aws
  class S3Service
    require 'aws-sdk-s3'
    PART_SIZE = 1024 * 1024 * 10

    def initialize(acl: InovaAwsS3.configuration.acl)
      @acl = acl
      @bucket_name = InovaAwsS3.configuration.bucket_name
      @region = InovaAwsS3.configuration.region
      @access_key = InovaAwsS3.configuration.access_key
      @secret_key = InovaAwsS3.configuration.secret_key
      @folder_name = InovaAwsS3.configuration.folder_name
      set_s3_client
      set_s3_resource
    end

    def set_s3_client
      @s3_client = Aws::S3::Client.new(region: @region, access_key_id: @access_key, secret_access_key: @secret_key)
    end

    def set_s3_resource
      @s3_resource = Aws::S3::Resource.new(region: @region, access_key_id: @access_key, secret_access_key: @secret_key)
    end

    def multipart_upload(filename: nil, file: nil, acl: "public-read")
      logger ||= Logger.new("#{Rails.root}/log/uploads.log")
      key = @folder_name + filename
      resp = @s3_client.create_multipart_upload({
                                                  bucket: @bucket_name,
                                                  key: key, acl: acl })
      current_part = 1
      parts = []
      file = File.open(file, 'rb')
      index = 1
      until file.eof?
        logger.info("#{key} Part #{current_part}")
        part = file.read(PART_SIZE)
        resp_upload = @s3_client.upload_part({
                                               body: part,
                                               bucket: @bucket_name,
                                               key: key,
                                               part_number: current_part,
                                               upload_id: resp.upload_id,
                                             })

        parts << { etag: resp_upload.etag, part_number: current_part }
        current_part = current_part + 1
      end
      resp_complete = @s3_client.complete_multipart_upload({
                                                             bucket: @bucket_name, # required
                                                             key: key, # required
                                                             multipart_upload: {
                                                               parts: parts
                                                             },
                                                             upload_id: resp.upload_id # required
                                                           })
      file.close()
      set_object(key).public_url
    end

    def onepart_upload(file_name: nil, tempfile: nil)
      file_name = "#{SecureRandom.urlsafe_base64}_#{Time.now.strftime('%s')}_#{File.extname(file_name)}".gsub(/\s+/, "")
      key = @folder_name + file_name
      obj = set_object(key)
      obj.upload_file(tempfile, { acl: @acl })
      obj.public_url
    end

    def set_object(key)
      @s3_resource.bucket(@bucket_name).object(key)
    end

    def change_acl(url: nil, new_acl: 'authenticated-read')
      return if url.nil?
      key = get_key_from_url(url)
      @s3_client.put_object_acl({ acl: new_acl, bucket: @bucket_name, key: key })
    rescue Exception
      puts "error in #{url}"
    end

    def get_key_from_url(url)
      # uri = URI.parse(url)
      # File.basename(uri.path)
      arr = url.split("/")
      arr[3..-1]&.join("/")
    end

    def valid_s3_url?(url)
      arr = url.split("/")
      host_name = "#{@bucket_name}.s3.#{@region}.amazonaws.com"
      return arr[2] == host_name
    end

    def get_key(filename)
      key = @folder_name + filename
      return key
    end

    def generate_get_presigned(url)
      key = get_key_from_url(url)
      object = @s3_resource.bucket(@bucket_name).object(key)
      presigned_url = object.presigned_url(:get, expires_in: 10800) # 3 hours
      presigned_url
    end

    def get_presigned_url(file_name)
      object = @s3_resource.bucket(@bucket_name).object(@folder_name.to_s + file_name)
      presigned_url = object.presigned_url(:put, { acl: "public-read" }) #, expires: 10*60)
      { presigned_url: presigned_url, file_url: object.public_url }
    end

    def delete_file(url)
      key = get_key_from_url(url)
      return if key.blank?
      @s3_client.delete_object({ bucket: @bucket_name, key: key })
    end
  end
end
